rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala

```diff
@@ -0,0 +1,294 @@
+
+  def subOrDec(exp: EVar)(implicit env: Env[Par]): Either[EVar, Par] =
+    subOrDec(exp.v) match {
+      case Left(varue) => Left(EVar(varue))
+      case Right(par)  => Right(par)
+    }
+
+  def substitute(term: Channel)(implicit env: Env[Par]): Channel =
+    term match {
+      case Quote(p) => Quote(substitute(p))
+      case ChanVar(v) =>
+        subOrDec(v) match {
+          case Left(_v) => ChanVar(_v)
+          case Right(p) => Quote(p)
+        }
+    }
+
+  def substitute(term: Par)(implicit env: Env[Par]): Par = {
+
+    def subExp(expxs: List[Expr]): Par =
+      (Par() /: expxs) { (par, expr) =>
+        expr match {
+          case e @ EVar(_) =>
+            subOrDec(e) match {
+              case Left(_e)    => par :+ _e
+              case Right(_par) => par ++ _par
+            }
+          case e @ _ => par :+ substitute(e)
+        }
+      }
+
+    val sends   = for { s <- term.sends } yield substitute(s)
+    val recvs   = for { r <- term.receives } yield substitute(r)
+    val evals   = for { e <- term.evals } yield substitute(e)
+    val news    = for { n <- term.news } yield substitute(n)
+    val matches = for { m <- term.matches } yield substitute(m)
+
+    val freeCount: Int =
+      (0 /: sends)((i, s) => i + s.freeCount) +
+        (0 /: recvs)((i, r) => i + r.freeCount) +
+        (0 /: evals)((i, e) => i + e.freeCount) +
+        (0 /: news)((i, n) => i + n.freeCount) +
+        (0 /: matches)((i, m) => i + m.freeCount)
+
+    Par(
+      sends,
+      recvs,
+      evals,
+      news,
+      Nil,
+      matches,
+      term.id,
+      freeCount
+    ) ++ subExp(term.exprs)
+  }
+
+  def substitute(term: Send)(implicit env: Env[Par]): Send = {
+    val _data = term.data map { par =>
+      substitute(par)
+    }
+    val freeCount = term.chan.freeCount + (0 /: _data) { (i, par) =>
+      i + par.freeCount
+    }
+    Send(
+      substitute(term.chan),
+      _data,
+      term.persistent,
+      freeCount
+    )
+  }
+
+  def substitute(term: Receive)(implicit env: Env[Par]): Receive = {
+    val _binds = for { (xs, chan) <- term.binds } yield {
+      (xs, substitute(chan))
+    }
+    val _body = substitute(term.body)
+    val freeCount = _body.freeCount + (0 /: _binds) {
+      case (i, (chanxs, chan)) =>
+        i + chan.freeCount + (0 /: chanxs) { (j, _chan) =>
+          j + _chan.freeCount
+        }
+    }
+    Receive(
+      _binds,
+      _body,
+      term.persistent,
+      _binds.size,
+      freeCount
+    )
+  }
+
+  def substitute(term: Eval)(implicit env: Env[Par]): Eval =
+    Eval(substitute(term.channel))
+
+  def substitute(term: New)(implicit env: Env[Par]): New =
+    New(term.bindCount, substitute(term.p))
+
+  def substitute(term: Match)(implicit env: Env[Par]): Match = {
+    val _value = substitute(term.value)
+    val _cases = for { (_case, par) <- term.cases } yield {
+      (_case, substitute(par))
+    }
+    Match(
+      _value,
+      _cases,
+      _value.freeCount +
```

Comment:
Same here.

Reply:
As an alternative, you should see that since you only ever replace bound variables, the free count will never change.